Add VidVerification payload generation to CHIPCryptoPAL

tcarmelveilleux opened the pull request:
#### Changes

- Adds necessary crypto primitives for VID Verification, which will be used by later PRS
- Adds the necessary unit tests based on spec test vectors.

#### Testing

- Added unit test coverage for the new code, matching independent test vectors in spec.

tcarmelveilleux committed:
Add VidVerification payload generation to CHIPCryptoPAL

This PR:

- Adds necessary crypto primitives for VID Verification,
  which will be used by later PRS
- Adds the necessary unit tests based on spec test vectors.

Testing done:

- Added unit test coverage for the new code, matching
  independent test vectors in spec.

restyled-commits committed:
Restyled by clang-format

andy31415 commented on src/crypto/CHIPCryptoPAL.h:
```diff
@@ -116,6 +117,30 @@ inline constexpr char kPIDPrefixForCNEncoding[]    = "Mpid:";
 inline constexpr size_t kVIDandPIDHexLength        = sizeof(uint16_t) * 2;
 inline constexpr size_t kMax_CommonNameAttr_Length = 64;
 
+inline constexpr uint8_t kFabricBindingVersionV1 = 1u;
```
could this be an enum class? That way we do not pass in raw values.

andy31415 replied:
It also prevents passing in invalid values and we would not need to do some extra checks (except on callers if they read some header)

andy31415 commented on src/crypto/CHIPCryptoPAL.cpp:
```diff
@@ -922,6 +922,91 @@ CHIP_ERROR DeriveGroupOperationalCredentials(const ByteSpan & epoch_key, const B
+
+    VerifyOrReturnError(outputSpan.size() >= kVendorFabricBindingMessageV1MaxSize, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    Encoding::BigEndian::BufferWriter writer(outputSpan.data(), outputSpan.size());
```
```suggestion
    Encoding::BigEndian::BufferWriter writer(outputSpan);
```

andy31415 commented on src/crypto/CHIPCryptoPAL.cpp:
```diff
@@ -922,6 +922,91 @@ CHIP_ERROR DeriveGroupOperationalCredentials(const ByteSpan & epoch_key, const B
+
+    VerifyOrReturnError(outputSpan.size() >= kVendorIdVerificationTbsV1MaxSize, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    Encoding::BigEndian::BufferWriter writer(outputSpan.data(), outputSpan.size());
```
```suggestion
    Encoding::BigEndian::BufferWriter writer(outputSpan);
```

andy31415 commented on src/crypto/CHIPCryptoPAL.cpp:
```diff
@@ -922,6 +922,91 @@ CHIP_ERROR DeriveGroupOperationalCredentials(const ByteSpan & epoch_key, const B
+ * @param[inout] outputSpan - Span that will receive the binding message. Must be large enough for the
+ *                            payload (otherwise CHIP_ERROR_BUFFER_TOO_SMALL) and will be resized to fit.
+ * @return CHIP_NO_ERROR on success, otherwise another CHIP_ERROR value representative of the failure.
+ */
```
These comments duplicate comments in the header. Let's delete them from here to have a single source of truth.

tcarmelveilleux replied:
Good catch. I had them to help me write the body.

andy31415 commented on src/crypto/CHIPCryptoPAL.cpp:
```diff
@@ -922,6 +922,91 @@ CHIP_ERROR DeriveGroupOperationalCredentials(const ByteSpan & epoch_key, const B
+    uint8_t fabricBindingVersion = vendorFabricBindingMessage[0];
+    VerifyOrReturnError(fabricBindingVersion == kFabricBindingVersionV1, CHIP_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrReturnError(outputSpan.size() >= kVendorIdVerificationTbsV1MaxSize, CHIP_ERROR_BUFFER_TOO_SMALL);
```
do we need this? Would `fit` not already catch this?

tcarmelveilleux replied:
I think that would work. I was trying to avoid touching buffer at all if too small.

andy31415 commented on src/crypto/CHIPCryptoPAL.cpp:
```diff
@@ -922,6 +922,91 @@ CHIP_ERROR DeriveGroupOperationalCredentials(const ByteSpan & epoch_key, const B
+    // Only V1 supported yet.
+    VerifyOrReturnError(fabricBindingVersion == kFabricBindingVersionV1, CHIP_ERROR_INVALID_ARGUMENT);
+
+    VerifyOrReturnError(outputSpan.size() >= kVendorFabricBindingMessageV1MaxSize, CHIP_ERROR_BUFFER_TOO_SMALL);
```
fit for writer should catch this right? Is there a need for early checks? Not having this would result in smaller code.

andy31415 commented on src/crypto/CHIPCryptoPAL.cpp:
```diff
@@ -922,6 +922,91 @@ CHIP_ERROR DeriveGroupOperationalCredentials(const ByteSpan & epoch_key, const B
+    writer.Put8(fabricBindingVersion).Put(rootPublicKey.ConstBytes(), rootPublicKey.Length()).Put64(fabricId).Put16(vendorId);
+
+    size_t actuallyWritten = 0;
+    if (!writer.Fit(actuallyWritten))
```
maybe this is also readable:

```cpp
VerifyOrReturnError(writer.Fit(actuallyWritten), CHIP_ERROR_BUFFER_TOO_SMALL);
```

Did not add as suggestion since unsure if this is really more readable. Seems more consistent with verify macros above.

balducci-apple commented on src/crypto/CHIPCryptoPAL.h:
```diff
@@ -116,6 +117,30 @@ inline constexpr char kPIDPrefixForCNEncoding[]    = "Mpid:";
+inline constexpr size_t kVendorIdVerificationClientChallengeSize = 32u;
+
+// VIDVerificationStatement := statement_version || vid_verification_signer_skid || vid_verification_statement_signature
+inline constexpr size_t kVendorIdVerificationStatementV1MaxSize =
```
Nit: Should we drop `Max` from these names ? afaik these are the _exact_ sizes.

tcarmelveilleux replied:
Fir some of them it's max, some it's exact. I will clarify.

tcarmelveilleux replied:
Done

balducci-apple reviewed (approved):
Nit: I'm not sure where else to put these functions but is CHIPCryptoPAL.h the appropriate place ? It does seem like CHIPCryptoPAL.h has some "crypto-adjacent" utilities so maybe it's fine.

tcarmelveilleux committed:
Fix unit test static analyzer warning

Error was:
```
../../src/crypto/tests/TestChipCryptoPAL.cpp:1573:38: note: Access of 'csr' at negative byte offset -2
 1573 |         csr[length - 2] = (uint8_t) (csr[length - 2] + 1);
      |                                      ^~~~~~~~~~~~~~~

WARNING TIDY ../../src/crypto/tests/TestChipCryptoPAL.cpp: 1 warning treated as error
WARNING Tidy ../../src/crypto/tests/TestChipCryptoPAL.cpp ended with code 1
```

Cause was that a EXPECT was used where ASSERT should be used, which leads to
an access in mbedTLS unit test version, that should not have run.

tcarmelveilleux committed:
Apply review comments

restyled-commits committed:
Restyled by clang-format

tcarmelveilleux committed:
Merge remote-tracking branch 'upstream/master' into add-vid-verify-crypto

tcarmelveilleux committed:
Disable a non-applicable clang-tidy
